libdnf/rpm/nevra.cpp

```diff
@@ -47,6 +47,205 @@ const std::vector<Nevra::Form> & Nevra::get_default_pkg_spec_forms() {
     return PKG_SPEC_FORMS;
 }
 
+std::vector<Nevra> Nevra::parse_all(const std::string & nevra_str, const std::vector<Form> & forms) {
+    std::vector<Nevra> result;
+    const char * evr_delim = nullptr;
+    const char * epoch_delim = nullptr;
+    const char * release_delim = nullptr;
+    const char * first_dot = nullptr;
+    const char * arch_delim = nullptr;
+    const char * end;
+
+    // parse nevra
+    const char * nevra_pattern = nevra_str.c_str();
+    for (end = nevra_pattern; *end != '\0'; ++end) {
+        if (*end == '-') {
+            evr_delim = release_delim;
+            release_delim = end;
+        } else if (*end == '.') {
+            arch_delim = end;
+            if (first_dot == nullptr) {
+                first_dot = end;
+            }
+        } else if (*end == ':') {
+            // ':' can be only once in nevra
+            if (epoch_delim != nullptr) {
+                return result;
+            }
+            epoch_delim = end;
+        } else if (*end == '(' || *end == '/' || *end == '=' || *end == '<' || *end == '>' || *end == ' ') {
+            return result;
+        }
+    }
+    for (auto form : forms) {
+        Nevra nevra;
+        switch (form) {
+            case Form::NEVRA: {
+                if (evr_delim == nullptr || release_delim == nullptr || arch_delim == nullptr) {
+                    continue;
+                }
+                const char * evr_delim_for_nevra = evr_delim;
+                const char * release_delim_for_nevra = release_delim;
+                const char * arch_delim_for_nevra = arch_delim;
+
+                // test name presence
+                if (evr_delim_for_nevra == nevra_pattern) {
+                    continue;
+                }
+
+                nevra.name.append(nevra_pattern, evr_delim_for_nevra);
```

Comment:
Why are you using `append` for assigning new value? Yes append works because the target variable is empty.
You may have overlooked the existence of the `assign` method, so you're using `append` instead.
Simply replace `append` with `assign`.
`nevra.name.assign(nevra_pattern, evr_delim_for_nevra);`

It is repeated in several places.

Reply:
I used append, because I was sure that the original string was empty. Changed to `assign`.